### ISSUE TITLE
[RFC]: Iterative Activation Processing

### ISSUE THREAD

**Issue by evanchueng** [vllm-project/vllm-omni#5493]
### Motivation.

# SUB RFC-3: Iterative Activation Processing

## Motivation

Data Parallelism (DP) provides superior throughput characteristics—zero inter-rank communication and high GEMM utilization. However, DP's independent replicas mean each rank processes the full sequence with all attention heads and all MoE experts.

### Attention Bottleneck

In standard DP execution, QKV projection is computed for all attention heads at once:

```python
# Standard DP: compute all heads at once
Q = linear(x)  # [seq_len, num_heads, head_dim] = [600K, 32, 128]
K = linear(x)  # [600K, 32, 128]
V = linear(x)  # [600K, 32, 128]
# QKV all resident in HBM: 14.4 GB

# Then apply LN + RoPE + FA all at once
# Peak memory = QKV (14.4GB) + LN/RoPE (14.4GB) + FA intermediates (~4.8GB) = ~33.6GB
```

### MoE Bottleneck

For MoE layers, each token is routed to a subset of experts. However, in standard DP, all tokens are processed through the expert computation, generating large intermediate tensors:

```python
# Standard MoE: all tokens through all experts at once
expert_outputs = []
for expert in experts:
    # Each expert processes all tokens (or selected tokens)
    out = expert(hidden_states)  # [seq_len, hidden_dim]
    expert_outputs.append(out)
# All expert outputs resident simultaneously: seq_len × num_experts × hidden_dim
```

For `seq_len=600K`, `num_experts=8`, `hidden_dim=4096`, FP16:

```
MoE intermediates = 600K × 8 × 4096 × 2 = 39.3 GB
```

Combined with attention activation, this makes per-layer memory >40GB, severely limiting batch size and throughput.

## Proposed Change

We introduce iterative processing for both attention and MoE layers: instead of computing all heads or all tokens at once, we loop over heads (for attention) and chunks of tokens (for MoE), keeping only a small working set in HBM.

### Part 1: Per-Head Iteration for Attention

Instead of computing QKV for all heads at once, each rank iterates through heads one at a time:

```python
# Sub-RFC 3: Per-Head Iteration (Attention)
outputs = []
for head_idx in range(num_heads):  # iterate one head at a time
    # Compute QKV for current head only
    Q_head = Q[:, :, head_idx, :]    # [seq_len, 1, head_dim]
    K_head = K[:, :, head_idx, :]
    V_head = V[:, :, head_idx, :]
    # LN + RoPE + FA only process current head
    out_head = attention(Q_head, K_head, V_head)
    outputs.append(out_head)
    # Current head's intermediates freed after each iteration
return torch.cat(outputs, dim=-1)
```

Memory reduction: Only one head's QKV + intermediates are held at a time, instead of all heads simultaneously.

### Part 2: Per-Chunk Iteration for MoE

For MoE layers, we process tokens in chunks rather than all at once:

```python
# Sub-RFC 3: Per-Chunk Iteration (MoE)
chunk_size = 4096  # tunable parameter
outputs = []
for start in range(0, seq_len, chunk_size):
    chunk_hidden = hidden_states[start:start+chunk_size]  # [chunk_size, hidden_dim]
    # Route tokens in this chunk to experts
    expert_outputs = []
    for expert in experts:
        out = expert(chunk_hidden)  # [chunk_size, hidden_dim]
        expert_outputs.append(out)
    # Combine expert outputs for this chunk
    chunk_output = combine(expert_outputs)  # [chunk_size, hidden_dim]
    outputs.append(chunk_output)
    # Free chunk intermediates after each iteration
return torch.cat(outputs, dim=0)
```

Memory reduction: Only `chunk_size` tokens' intermediates are held at a time, instead of full `seq_len`.

### Attention (seqlen=600K, num_heads=32, head_dim=128, FP16)

| Component | Standard DP | Per-Head Iteration |
|-----------|------------|-------------------|
| QKV | 14.4 GB | 0.45 GB (1 head) |
| LN + RoPE | 14.4 GB | 0.45 GB (1 head) |
| FA intermediates | ~4.8 GB | ~0.15 GB (1 head) |
| **Peak** | **~33.6 GB** | **~1.05 GB** |

### MoE (seqlen=600K, num_experts=8, hidden_dim=4096, FP16)

| Configuration | Intermediates Memory |
|--------------|---------------------|
| Standard (full seq) | 600K × 8 × 4096 × 2 = 39.3 GB |
| Chunk size = 4096 | 4096 × 8 × 4096 × 2 = 268 MB |
| Chunk size = 8192 | 536 MB |
| Chunk size = 16384 | 1.07 GB |

### Combined per-layer peak

| Scenario | Attention | MoE | Total |
|----------|-----------|-----|-------|
| Standard DP | 33.6 GB | 39.3 GB | ~72.9 GB |
| RFC-3 (chunk=4096) | 1.05 GB | 0.27 GB | **~1.32 GB** |

**Result:** Activation memory per layer is reduced from ~73 GB to ~1.3 GB, a **~55× reduction**. No additional inter-rank communication (within-rank iteration only).

## Integration with Other RFCs

| RFC | Integration |
|-----|------------|
| Sub-RFC 1 (Weight Offload) | Weight offload frees HBM; iterative processing reduces activation memory. Together: HBM holds only 2 layers of weights (~3.2 GB) + 1 head/chunk of activations (~1.3 GB) ≈ 4.5 GB per layer. |
| Sub-RFC 2 (Async DP) | DP execution distributes prompts; iterative processing applies within each rank independently. |
| Sub-RFC 4 (Ref-Target KV) | Independent optimization; can be combined. Ref KV cache reduces cross-attention compute, iterative processing reduces activation memory. |

### QKV Projection
- Each head's QKV must be computed independently
- May require custom kernels or loop-based forward passes
- The QKV weight matrix can be sliced per-head: `W_qkv[head_idx]` → only current head's projection

### RoPE
- Applied per head; can be computed on the fly without full Q/K allocation
- `inv_freq` buffer is shared across heads (already small: ~0.5 MB)

### Flash Attention
- Works with per-head Q/K/V inputs
- FlashAttention-2 (CUDA) and CANN FlashAttention (NPU) both support single-head inputs
- No accuracy loss — mathematically equivalent to batch attention

### MoE Chunk Size
- Tunable parameter balancing memory vs. compute efficiency
- Larger chunks improve GEMM utilization but increase memory
- Recommended starting point: 4096 (268 MB intermediates)
- Can be auto-tuned based on available HBM (reported by `torch.cuda.mem_get_info` or NPU equivalent)

## Hardware Support

| Operation | NVIDIA GPU | Ascend NPU |
|-----------|-----------|------------|
| Per-head QKV slicing | CUDA | CANN |
| Per-head Flash Attention | FlashAttention-2 | CANN FlashAttention |
| Per-chunk MoE | CUDA | CANN |

## Summary

| Metric | Before (Standard DP) | After (RFC-3) |
|--------|---------------------|---------------|
| Attention peak | ~33.6 GB | ~1.05 GB |
| MoE peak | ~39.3 GB | ~0.27 GB (chunk=4096) |
| Total per layer | ~72.9 GB | ~1.32 GB |
| Reduction | — | ~55× |
| Additional communication | — | None (within-rank iteration) |

### Proposed Change.

Please provide the detailed design document of the RFC using the [template](https://docs.google.com/document/d/1jcgR3cDaUQH3VczD4ZcKaJAoYWHjCmnYzHkCNyz-9fk/edit?usp=sharing).

### CC List.

@hsliuustc0106 @Gaohan123 @david6666666 @bjf-frz 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.
